precise/iaf_psc_exp_ps_time_reversal.cpp

```diff
@@ -0,0 +1,720 @@
+    c3_ =(tau_m_ * tau_m_)/ (c_m_ * (tau_m_ - tau_ex_));
+    c4_ =tau_ex_/tau_m_;
+    c5_ =(c_m_ * U_th_)/tau_m_;
+    c6_ =1-(tau_ex_/tau_m_);
+
+    //parallel line
+
+    d1_ = tau_m_ *  c_m_ ;
+    d2_ = tau_m_ * tau_ex_;
+    d3_ = c_m_ * (tau_m_ - tau_ex_);
+}
+
+void nest::iaf_psc_exp_ps_time_reversal::Parameters_::get(DictionaryDatum & d) const
+{
+  def<double>(d, names::E_L, E_L_);
+  def<double>(d, names::I_e, I_e_);
+  def<double>(d, names::V_th, U_th_+E_L_);
+  def<double>(d, names::V_min, U_min_+E_L_);
+  def<double>(d, names::V_reset, U_reset_+E_L_);
+  def<double>(d, names::C_m, c_m_);
+  def<double>(d, names::tau_m, tau_m_);
+  def<double>(d, names::tau_syn_ex, tau_ex_);
+  def<double>(d, names::tau_syn_in, tau_in_);
+  def<double>(d, names::t_ref, t_ref_);
+
+}
+
+double nest::iaf_psc_exp_ps_time_reversal::Parameters_::set(const DictionaryDatum & d)
+{
+
+  updateValue<double>(d, names::tau_m, tau_m_);
+  updateValue<double>(d, names::tau_syn_ex, tau_ex_);
+  updateValue<double>(d, names::tau_syn_in, tau_in_);
+  updateValue<double>(d, names::C_m, c_m_);
+  updateValue<double>(d, names::t_ref, t_ref_);
+  updateValue<double>(d, names::I_e, I_e_);
+
+
+  // if U0_ is changed, we need to adjust all variables defined relative to U0_
+  const double ELold = E_L_;
+  updateValue<double>(d, names::E_L, E_L_);
+  const double delta_EL = E_L_ - ELold;
+
+  if(updateValue<double>(d, names::V_reset, U_reset_))
+    U_reset_ -= E_L_;
+  else
+    U_reset_ -= delta_EL;
+
+  if (updateValue<double>(d, names::V_th, U_th_))
+    U_th_ -= E_L_;
+  else
+    U_th_ -= delta_EL;
+
+  if (updateValue<double>(d, names::V_min, U_min_))
+    U_min_ -= E_L_;
+  else
+    U_min_ -= delta_EL;
+  
+  if ( U_reset_ >= U_th_ )
+    throw BadProperty("Reset potential must be smaller than threshold.");
+  
+  if ( U_reset_ < U_min_ )
+    throw BadProperty("Reset potential must be greater equal minimum potential.");
+  
+  if ( c_m_ <= 0 )
+    throw BadProperty("Capacitance must be strictly positive.");
+  
+  if ( t_ref_ < 0 )
+    throw BadProperty("Refractory time must not be negative.");
+  
+  if ( tau_m_ <= 0 || tau_ex_ <= 0 || tau_in_ <= 0 )
+    throw BadProperty("All time constants must be strictly positive.");
+
+   if ( tau_m_ == tau_ex_ || tau_m_ == tau_in_ )
+    throw BadProperty("Membrane and synapse time constant(s) must differ."
+                      "See note in documentation.");
+
+   calc_const_spike_test_();
+
+   return delta_EL;
+}
+
+void nest::iaf_psc_exp_ps_time_reversal::State_::get(DictionaryDatum & d, 
+                                       const Parameters_ & p) const
+{
+  def<double>(d, names::V_m, y2_ + p.E_L_);  // Membrane potential
+  def<bool>(d, names::is_refractory, is_refractory_);
+  def<double>(d, names::t_spike, Time(Time::step(last_spike_step_)).get_ms());
+  def<double>(d, names::offset, last_spike_offset_);
+  def<double>(d, names::y1_ex, y1_ex_);
+  def<double>(d, names::y1_in, y1_in_); // y1 state
```

Comment:
If I understand the code correctly, these are the excitatory and inhibitory synaptic current. Please use the pre-defined names `I_syn_ex` and `I_syn_in`, respectively.